// @marblejs/core/dist/+internal/http/contentType.util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ContentType;
(function (ContentType) {
    ContentType["APPLICATION"] = "application/*";
    ContentType["APPLICATION_RTF"] = "application/rtf";
    ContentType["APPLICATION_ZIP"] = "application/zip";
    ContentType["APPLICATION_X_RAR"] = "application/x-rar-compressed";
    ContentType["APPLICATION_X_TAR"] = "application/x-tar";
    ContentType["APPLICATION_X_TZ_COMPRESSED"] = "application/x-7z-compressed";
    ContentType["APPLICATION_X_WWW_FORM_URLENCODED"] = "application/x-www-form-urlencoded";
    ContentType["APPLICATION_PDF"] = "application/pdf";
    ContentType["APPLICATION_JSON"] = "application/json";
    ContentType["APPLICATION_JAVASCRIPT"] = "application/javascript";
    ContentType["APPLICATION_ECMASCRIPT"] = "application/ecmascript";
    ContentType["APPLICATION_XML"] = "application/xml";
    ContentType["APPLICATION_OCTET_STREAM"] = "application/octet-stream";
    ContentType["APPLICATION_VND_API_JSON"] = "application/vnd.api+json";
    ContentType["TEXT_PLAIN"] = "text/plain";
    ContentType["TEXT_HTML"] = "text/html";
    ContentType["TEXT_CSS"] = "text/css";
    ContentType["TEXT_CSV"] = "text/csv";
    ContentType["IMAGE_WEBP"] = "image/webp";
    ContentType["IMAGE_JPEG"] = "image/jpeg";
    ContentType["IMAGE_PNG"] = "image/png";
    ContentType["IMAGE_GIF"] = "image/gif";
    ContentType["IMAGE_TIFF"] = "image/tiff";
    ContentType["IMAGE_SVG_XML"] = "image/svg+xml";
    ContentType["AUDIO_MPEG"] = "audio/mpeg";
    ContentType["AUDIO_OGG"] = "audio/ogg";
    ContentType["AUDIO"] = "audio/*";
    ContentType["VIDEO_WEBM"] = "video/webm";
    ContentType["VIDEO_MP4"] = "video/mp4";
    ContentType["FONT_TTF"] = "font/ttf";
    ContentType["FONT_WOFF"] = "font/woff";
    ContentType["FONT_WOFF2"] = "font/woff2";
    ContentType["MULTIPART_FORM_DATA"] = "multipart/form-data";
})(ContentType = exports.ContentType || (exports.ContentType = {}));
exports.getContentType = (headers) => {
    const contentType = headers['content-type'] || headers['Content-Type'];
    return contentType
        ? Array.isArray(contentType) ? contentType[0] : String(contentType)
        : '';
};
