// @marblejs/core/dist/context/context.factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const R = require("fp-ts/lib/Reader");
const M = require("fp-ts/lib/Map");
const Ord_1 = require("fp-ts/lib/Ord");
const ordContextToken = Ord_1.contramap((t) => t._id, Ord_1.ordString);
const setoidContextToken = { equals: ordContextToken.equals };
const isReader = (x) => !!x.run;
exports.createContext = () => M.empty;
exports.register = (boundDependency) => (context) => M.insert(setoidContextToken)(boundDependency.token, isReader(boundDependency.dependency)
    ? boundDependency.dependency
    : boundDependency.dependency(context), context);
exports.registerAll = (boundDependencies) => (context) => boundDependencies.reduce((ctx, dependency) => exports.register(dependency)(ctx), context);
exports.lookup = (context) => (token) => M.lookup(ordContextToken)(token, context).map(dependency => {
    if (!isReader(dependency)) {
        return dependency;
    }
    const boostrapedDependency = dependency.run(context);
    context.set(token, boostrapedDependency);
    return boostrapedDependency;
});
exports.bindTo = (token) => (dependency) => ({ token, dependency });
exports.reader = R.ask().map(exports.lookup);
